### DISCUSSION TITLE
How to get IScheduler when using QuartzHostedService

### DISCUSSION THREAD

**Discussion by pdonovan** [quartznet/quartznet#1013] [Q&A] [answered]
I'm updating our code from 3.0.7 to 3.2.2 and am trying to simplify our Quartz scheduler construction to use the DI extensions, but I've hit a problem I'm not sure how to solve.

We have places in our system that call IScheduler.ScheduleJob() in response to messages that can occur at any time. When we create it ourselves we keep a reference to the IScheduler that StdSchedulerFactory.GetScheduler() returns. But QuartzHostedService is internal, so I can't use it in (for example) an ASP.NET Core Controller constructor. I've tried injecting Quartz.IScheduler, but the DI system says Quartz.IScheduler isn't registered.

Is there something I'm missing or is this not possible with these extensions?

**Comment by lahma** [comment, marked as answer]
I'm assuming that you have single scheduler running and thus single scheduler factory. You should be able to inject `ISchedulerFactory` to your controller and then call `await GetScheduler()` on it to get the active instance. `GetScheduler()` is async operation so easier to get when needed instead of injecting (you could also add it to container as scoped service if you want to).

If you at some point want to separate Quartz hosted service to separate process then you would probably want to configure your web scheduler separately with `ZeroSizeThreadPool` and not start it, this would allow just to do admin tasks.

**Reply by pdonovan** [reply]
Thanks Marko, that works.
